feat: add helper for event timestamps

rafaelugolini opened the pull request:
related to #30

sslivkoff commented:
looking good. a few more things then ready to merge:
- an `include_timestamps` input in `async_get_events` that if true will add a `timestamp` field to the output
- a similar `include_timestamps` input for a few functions in the codebase that use `async_get_events`. in each case simply pass the parameter to `async_get_events`. probably want in the following functions:
    - `evm.async_get_erc20_transfers`
    - `aave_utils.async_get_deposits` and `aave_utils.async_get_withdrawals`
    - `uniswap_v2_utils.async_get_pool_swaps`, `uniswap_v2_utils.async_get_pool_mints`, `uniswap_v2_utils.async_get_pool_burns`
    - `uniswap_v3_utils.async_get_pool_swaps`
    - `balancer_utils.async_get_pool_swaps`
- a very simple test of `async_get_events` that makes sure `include_timestamps=True` produces a correct output. like for example choose a random uniswap pool, get the swaps over a known hardcoded range of 10 blocks, and verify that the timestamps for those swaps are correct

default value for `include_timestamps` should be false in all cases because it will add extra work especially for large queries that span many blocks

rafaelugolini commented:
hey @sslivkoff, I added most of the things and a few extras; I can remove it if you don't think it fits there:

- I enhanced the CLI  to support timestamps
- Added a small example on README

I also have some questions:

1. Would it be useful to add `include_timestamps` on `src/ctc/protocols/ens_utils/registrar.py:async_get_registrations`? I tested that function I got an exception: ```ctc.spec.exceptions.rpc_exceptions.RpcException: RPC ERROR: query returned more than 10000 results```
2.  Would it be useful to add `include_timestamps` on `src/ctc/protocols/ens_utils/resolver.async_get_text_changes`? I think it could be nice but I would implement it differently to only add it over the returning rows(line 216)

sslivkoff commented on src/ctc/cli/commands/data/erc20/transfers_command.py:
```diff
@@ -57,5 +65,7 @@ async def async_transfers_command(
         end_block=end_block,
+        include_timestamps=timestamps,
     )
+    transfers = transfers.astype({'timestamp': 'str'})
```
actually why convert to str at all? isnt int timestamp preferable?

rafaelugolini replied:
@sslivkoff it's just to show a nicer output in the terminal

rafaelugolini replied:
with cast to `str`
```bash
(venv) ❯ ctc events 0x956f47f50a910163d8bf957cf5846d573e7f87ca Transfer --blocks "[14000000, 14000010]" --timestamps
got events 0x956f47f50a910163d8bf957cf5846d573e7f87ca 14000000 14000010
7
     block  │   timestamp  │                                        from  │                                          to  │                    value  
────────────┼──────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼───────────────────────────
  14000000  │  1642114795  │  0xbb2e5c2ff298fd96e166f90c8abacaf714df14f8  │  0x27239549dd40e1d60f5b80b0c4196923745b1fd2  │  58823967281975952181439  
  14000000  │  1642114795  │  0x27239549dd40e1d60f5b80b0c4196923745b1fd2  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  58823967281975952181439  
  14000003  │  1642114825  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  73435661712369422675474  
  14000005  │  1642114852  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9e2336aef4157944f201becd90ccb24e298660cb  │  14767355748583855822194  
  14000007  │  1642114881  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  25820219943704516282552  
  14000008  │  1642114895  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  36756489418238270476995  
  14000008  │  1642114895  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  46197251138161945267759  
  ```
without the cast
```bash
(venv) ❯ ctc events 0x956f47f50a910163d8bf957cf5846d573e7f87ca Transfer --blocks "[14000000, 14000010]" --timestamps
got events 0x956f47f50a910163d8bf957cf5846d573e7f87ca 14000000 14000010
7
     block  │      timestamp  │                                        from  │                                          to  │                    value  
────────────┼─────────────────┼──────────────────────────────────────────────┼──────────────────────────────────────────────┼───────────────────────────
  14000000  │  1,642,114,795  │  0xbb2e5c2ff298fd96e166f90c8abacaf714df14f8  │  0x27239549dd40e1d60f5b80b0c4196923745b1fd2  │  58823967281975952181439  
  14000000  │  1,642,114,795  │  0x27239549dd40e1d60f5b80b0c4196923745b1fd2  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  58823967281975952181439  
  14000003  │  1,642,114,825  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  73435661712369422675474  
  14000005  │  1,642,114,852  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9e2336aef4157944f201becd90ccb24e298660cb  │  14767355748583855822194  
  14000007  │  1,642,114,881  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  25820219943704516282552  
  14000008  │  1,642,114,895  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  36756489418238270476995  
  14000008  │  1,642,114,895  │  0x9928e4046d7c6513326ccea028cd3e7a91c7590a  │  0x94b0a3d511b6ecdb17ebf877278ab030acb0a878  │  46197251138161945267759  
```

sslivkoff replied:
oh that makes sense for display purposes. in that case it would be good to only do it `if output is None` so that it doesnt affect csv outputs

sslivkoff replied:
oh that makes sense for display purposes. in that case it would be good to only do it if output is None so that it doesnt affect csv outputs

sslivkoff reviewed (changes requested):
looks good. added a couple comments

i think you can skip the ens files for now because they need to be refactored anyway. they have some inefficiencies give include_timestamps some problems

sslivkoff commented on src/ctc/evm/event_utils/event_crud.py:
```diff
@@ -219,3 +219,16 @@ async def async_download_events(
     )
+
+
+async def async_get_event_timestamps(events: spec.DataFrame) -> spec.DataFrame:
```
return type should be typing.Sequence[int] not dataframe

sslivkoff commented on src/ctc/cli/commands/data/events_command.py:
```diff
@@ -104,15 +114,20 @@ async def async_events_command(
+                for column in events.columns
+                if column.startswith('arg__')
             ]
+            if timestamps:
+                columns.insert(0, 'timestamp')
+                events = events.astype({'timestamp': 'str'})
```
this snipped that inserts the field should be moved to evm.event_utils.event_crud.async_get_events instead of here

that way it can be used not just by the cli interface but also all the python functions

rafaelugolini replied:
@sslivkoff I added the snipped into `evm.event_utils.event_crud.async_get_events` but I also need something here because the code only display the index + columns starting with `arg__`. is it ok with you?

sslivkoff reviewed (changes requested):
just ran the tests. some things are failing because the timestamp field insertion is only in the cli command implementation instead of the upstream evm.event_utils.event_crud.async_get_events 

I'll throw together a CONTRIBUTING.md doc to make it easier to run the tests

sslivkoff commented:
for the relevant test commands see https://github.com/fei-protocol/checkthechain/blob/main/CONTRIBUTING.md

rafaelugolini commented:
@sslivkoff cool, thanks for the feedback. I will work on this over the weekend.

rafaelugolini commented:
@sslivkoff btw... I did a rebase and I fix all the issues with the tests on the cli but I don't know if you like the it or if you want it in a separate PR:

```patch
diff --git a/pyproject.toml b/pyproject.toml
index 7baccbb..147ec34 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -61,6 +61,7 @@ dependencies = [
     'pycryptodome >=3.9.1, <4',  # for keccak()
     'eth_abi_lite >=3.0.3, <4',  # for abi encoding
     'idna >=3.3, <4',  # ENS resolution
+    'rlp >= 3.0.0', # create address
 ]
 
 [project.optional-dependencies]
diff --git a/src/ctc/cli/commands/admin/alias_command.py b/src/ctc/cli/commands/admin/alias_command.py
index 6a8a315..992863d 100644
--- a/src/ctc/cli/commands/admin/alias_command.py
+++ b/src/ctc/cli/commands/admin/alias_command.py
@@ -47,7 +47,7 @@ def get_command_spec() -> toolcli.CommandSpec:
                 'action': 'store_true',
             },
         ],
-        'examples': ['', '--status', '--append'],
+        'examples': ['', '--status'],
     }
 
 
diff --git a/src/ctc/evm/abi_utils/abi_summary.py b/src/ctc/evm/abi_utils/abi_summary.py
index e9106fa..07c858a 100644
--- a/src/ctc/evm/abi_utils/abi_summary.py
+++ b/src/ctc/evm/abi_utils/abi_summary.py
@@ -157,9 +157,7 @@ def summarize_contract_abi_functions(
     has_receive = any(item['type'] == 'receive' for item in contract_abi)
     has_fallback = any(item['type'] == 'fallback' for item in contract_abi)
 
-    toolstr.print_text_box(
-        'Special Functions Present', style=styles['title'], indent=None
-    )
+    toolstr.print_text_box('Special Functions Present', style=styles['title'])
     print()
     special_rows = [
         ('constructor', has_constructor),
```

```bash
======================================================================== 300 passed in 83.09s (0:01:23) =========================================================================
````

sslivkoff reviewed (approved):
looking good

sslivkoff commented:
excellent stuff. thanks for the PR

sslivkoff commented:
> @sslivkoff btw... I did a rebase and I fix all the issues with the tests on the cli but I don't know if you like the it or if you want it in a separate PR:
> 
> ```diff
> diff --git a/pyproject.toml b/pyproject.toml
> index 7baccbb..147ec34 100644
> --- a/pyproject.toml
> +++ b/pyproject.toml
> @@ -61,6 +61,7 @@ dependencies = [
>      'pycryptodome >=3.9.1, <4',  # for keccak()
>      'eth_abi_lite >=3.0.3, <4',  # for abi encoding
>      'idna >=3.3, <4',  # ENS resolution
> +    'rlp >= 3.0.0', # create address
>  ]
>  
>  [project.optional-dependencies]
> diff --git a/src/ctc/cli/commands/admin/alias_command.py b/src/ctc/cli/commands/admin/alias_command.py
> index 6a8a315..992863d 100644
> --- a/src/ctc/cli/commands/admin/alias_command.py
> +++ b/src/ctc/cli/commands/admin/alias_command.py
> @@ -47,7 +47,7 @@ def get_command_spec() -> toolcli.CommandSpec:
>                  'action': 'store_true',
>              },
>          ],
> -        'examples': ['', '--status', '--append'],
> +        'examples': ['', '--status'],
>      }
>  
>  
> diff --git a/src/ctc/evm/abi_utils/abi_summary.py b/src/ctc/evm/abi_utils/abi_summary.py
> index e9106fa..07c858a 100644
> --- a/src/ctc/evm/abi_utils/abi_summary.py
> +++ b/src/ctc/evm/abi_utils/abi_summary.py
> @@ -157,9 +157,7 @@ def summarize_contract_abi_functions(
>      has_receive = any(item['type'] == 'receive' for item in contract_abi)
>      has_fallback = any(item['type'] == 'fallback' for item in contract_abi)
>  
> -    toolstr.print_text_box(
> -        'Special Functions Present', style=styles['title'], indent=None
> -    )
> +    toolstr.print_text_box('Special Functions Present', style=styles['title'])
>      print()
>      special_rows = [
>          ('constructor', has_constructor),
> ```
> 
> ```shell
> ======================================================================== 300 passed in 83.09s (0:01:23) =========================================================================
> ```

these all look reasonable. can submit as separate pr

for the rlp tho, want to leave that out of the main dependencies. can put it in the test depedencies. want as few dependencies as possible. leaving it in as an optional dependency is fine for those that want rlp stuff but I think theres also a simple way to implement the rlp encoding needed for creating addresses without needing a full rlp lib
